Server/JokesOnYou.Web.Api/Controllers/UserController.cs

```diff
@@ -1,48 +1,143 @@
-﻿using JokesOnYou.Web.Api.DTOs;
+﻿using AutoMapper;
+using JokesOnYou.Web.Api.DTOs;
+using JokesOnYou.Web.Api.Models;
 using JokesOnYou.Web.Api.Services.Interfaces;
 using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
 using System;
 using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
 using System.Linq;
 using System.Threading.Tasks;
 
 namespace JokesOnYou.Web.Api.Controllers
 {
-    [Route("api/[controller]")]
-    [ApiController]
     [Authorize]
+    [Route("[controller]")]
     public class UserController : ControllerBase
     {
-        private readonly IUserService _userService;
 
-        public UserController(IUserService userService)
+        readonly IUserService _userService;
+        readonly ILogger<UserController> _logger;
+        readonly IMapper _mapper;   
+        readonly ITokenService _tokenService;
+        public UserController(IUserService userService, ILogger<UserController> logger, IMapper mapper, ITokenService tokenService)
         {
+            _logger = logger;
+            _mapper = mapper;
             _userService = userService;
+            _tokenService = tokenService;
         }
 
+        //ok so returning user is dumb, since we return sensitive data, we need better DTO system, Valve plz fix
         [HttpGet]
-        public async Task<ActionResult> GetUsers()
+        public ActionResult<IEnumerable<User>> GetUsers()
         {
-            throw new NotImplementedException();
+            return Ok(_userService.GetAll());
         }
 
+        //idk if its from body or Http get thing
         [HttpGet("{id}")]
-        public async Task<ActionResult> GetUser(string id)
+        public  ActionResult<User> GetUserById(string id)
         {
-            throw new NotImplementedException();
-        }
+            var user =  _userService.GetUserById(id);
 
-        [HttpPatch]
-        public async Task<ActionResult> UpdateUser(UserUpdateDTO userUpdateDTO)
+            if (user != null)
+            {
+                return user;
+            }
+            _logger.LogInformation($"User not found; id: {id}");
+            return NotFound();
+            
+        }
+        [HttpPut("{id}")]
+        public async Task<ActionResult> UpdateUser(string id, [FromBody]UserUpdateDTO userUpdateDTO)
         {
-            throw new NotImplementedException();
+            var user = _mapper.Map<User>(userUpdateDTO);
+
+
+            if (await _userService.UpdateUser(user))
+            {
+                return Ok();
+            }
+            _logger.LogInformation($"Something went wrong with updating user that had id {user.Id}");
+
+            return NotFound();
         }
 
+        [AllowAnonymous]
+        [HttpPost("authenticate")]
+        public async Task<ActionResult> Authenticate([FromBody] UserLoginDTO userDto)
+        {
+            User user;
+            if (string.IsNullOrWhiteSpace(userDto.Password))
```

Comment:
If we add [required] on to our logindto then we wont even reach this far because the model would be invalid since the api makes sure of that

Reply:
i think there will be a stop at the website for not sending the info if its empty :P but yes adding required is good!

Reply:
We will have client validation dw

Reply:
So remove every `string.IsNullOrWhiteSpace(userDto.Parameter)`?

Reply:
yes, we dont need to validate the incoming objects, data annotations are enough